### PR TITLE
Point plus extended

### PR DESCRIPTION
**Pull request by nkhadka21**
added a point + extended source skeleton.

### CODE REVIEW

File: slsim/Sources/point_plus_extended_source.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import numpy.random as random
+from slsim.Sources.source_pop_base import SourcePopBase
+import numpy as np
+from slsim.selection import deflector_cut
+from slsim.Sources.galaxies import galaxy_projected_eccentricity
+
+
+class PESource(SourcePopBase):
```
Comment (sibirrer) [inline slsim/Sources/point_plus_extended_source.py:8]: I would use a more meaningful class name

File: slsim/Sources/point_plus_extended_source.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        pes_list,
+        cosmo,
+        sky_area,
+        kwargs_cut,
+        variability_model=None,
+        kwargs_variability_model=None,
+    ):
+        """
+
+        :param pes_list: list of dictionary with point and extended source parameters
```
Comment (sibirrer) [inline slsim/Sources/point_plus_extended_source.py:22]: pes_list is hard to interpret as a parameter name

File: slsim/Sources/point_plus_extended_source.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        super(PESource, self).__init__(
+            cosmo=cosmo,
+            sky_area=sky_area,
+            variability_model=variability_model,
+            kwargs_variability_model=kwargs_variability_model,
+        )
+        self.n = len(pes_list)
+        ## check missing kewords in astropy table.
+        column_names = pes_list.colnames
+        if "ellipticity" not in column_names:
```
Comment (sibirrer) [inline slsim/Sources/point_plus_extended_source.py:44]: these lines of code are perhaps very similar/duplicate to the one in the extended source class. Perhaps we can mere them, or even have this class inherited from the extended source class?

File: slsim/Sources/point_plus_extended_source.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    def draw_source(self):
+        """Choose source at random.
+
+        :return: dictionary of source
+        """
+
+        index = random.randint(0, self._num_select - 1)
+        point_extended_source = self._point_extended_select[index]
+
+        if point_extended_source["e1"] == -1 or point_extended_source["e2"] == -1:
```
Comment (sibirrer) [inline slsim/Sources/point_plus_extended_source.py:73]: these lines of code are perhaps very similar/duplicate to the one in the extended source class. Perhaps we can mere them, or even have this class inherited from the extended source class?

File: slsim/Sources/source.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-from slsim.Sources.source_variability.variability import (
+from slsim.Sources.Source_variability.variability import (
```
Comment (sibirrer) [inline slsim/Sources/source.py:1]: can you rename this folder to SourceVariability?

File: slsim/Sources/source.py [hunk 6]
```diff
@@ -90,3 +90,19 @@ def magnitude(self, band, image_observation_times=None):
         else:
             source_mag = self.source_dict[band_string]
             return source_mag
+
+    def es_magnitude(self, band):
```
Comment (sibirrer) [inline slsim/Sources/source.py:94]: I would rename the definition to something better interpretable. 'es' is hard to understand. You can also write it out as extended_source_magnitude()

File: slsim/Util/param_util.py [hunk 7]
```diff
@@ -57,6 +57,25 @@ def random_ra_dec(ra_min, ra_max, dec_min, dec_max, n):
     return ra, dec
 
 
+def random_radec_string(ra_min, ra_max, dec_min, dec_max, n):
+    """Generates n number of random "ra, dec" string with in a given limits.
```
Comment (sibirrer) [inline slsim/Util/param_util.py:61]: wording 'with in a given limits' -> within given limits

File: slsim/image_simulation.py [hunk 8]
```diff
@@ -266,35 +279,41 @@ def point_source_image_without_variability(
...
+            delta_pix=delta_pix,
+            num_pix=num_pix,
+            transform_pix2angle=transform_pix2angle,
+        )
+        ra_image_values = image_data["ra_image"]
+        dec_image_values = image_data["dec_image"]
+        magnitude = lens_class.point_source_magnitude(band, lensed=True)
+        amp = magnitude_to_amplitude(magnitude, mag_zero_point)
+    else:
+        ra_image_values = np.zeros(1)
```
Comment (sibirrer) [inline slsim/image_simulation.py:309]: is this desired? Perhaps you can have empty lists instead?

File: slsim/image_simulation.py [hunk 9]
```diff
@@ -323,41 +342,49 @@ def point_source_image_at_time(
...
-    image_data = point_source_coordinate_properties(
-        lens_class=lens_class,
-        band=band,
-        mag_zero_point=mag_zero_point,
-        delta_pix=delta_pix,
-        num_pix=num_pix,
-        transform_pix2angle=transform_pix2angle,
-    )
-
+    kwargs_model, kwargs_params = lens_class.lenstronomy_kwargs(band="i")
```
Comment (sibirrer) [inline slsim/image_simulation.py:352]: this is a hard-coded i-band. Is this needed? If you do not need a band, I would not require you to have a certain band be present

File: slsim/image_simulation.py [hunk 10]
```diff
@@ -323,41 +342,49 @@ def point_source_image_at_time(
...
+            transform_pix2angle=transform_pix2angle,
+        )
+        ra_image_values = image_data["ra_image"]
+        dec_image_values = image_data["dec_image"]
+        variable_mag = lens_class.point_source_magnitude(
+            band=band, lensed=True, time=time
+        )
+        variable_amp = magnitude_to_amplitude(variable_mag, mag_zero_point)
+    else:
+        ra_image_values = np.zeros(1)
```
Comment (sibirrer) [inline slsim/image_simulation.py:376]: is this desired this way? Perhaps better to return empty lists?

File: slsim/image_simulation.py [hunk 11]
```diff
@@ -395,10 +424,187 @@ def point_source_image_with_variability(
...
+            lens_class=lens_class,
+            band=band,
+            mag_zero_point=mag_zero_point,
+            delta_pix=delta_pix,
+            num_pix=num_pix,
+            psf_kernel=psf_kernel,
+            transform_pix2angle=transform_pix2angle,
+            time=t_obs,
+        )
+    image = convolved_deflector_source + sum(image_ps)
```
Comment (sibirrer) [inline slsim/image_simulation.py:557]: are you sure this is doing the right thing? What is this sum over?

File: slsim/lens.py [hunk 12]
```diff
@@ -360,8 +364,8 @@ def point_source_magnitude(self, band, lensed=False, time=None):
         :type band: string
         :param lensed: if True, returns the lensed magnified magnitude
         :type lensed: bool
-        :param time: time is a image observation time which is a astropy.unit object. If
-            None, provides magnitude without variability.
+        :param time: time is a image observation time in day. If None, provides
```
Comment (sibirrer) [inline slsim/lens.py:367]: 'in units of days'

**Review by sibirrer — CHANGES_REQUESTED** [review]
Looks overall good. I have some minor requests in the comments

File: slsim/Sources/point_plus_extended_source.py (hunk 2 above)
Comment (nkhadka21) [reply slsim/Sources/point_plus_extended_source.py:22]: Renamed it as point_plus_extended_source_list.

File: slsim/Sources/point_plus_extended_source.py (hunk 4 above)
Comment (nkhadka21) [reply slsim/Sources/point_plus_extended_source.py:73]: Inherited from Galaxies class.

File: slsim/image_simulation.py (hunk 8 above)
Comment (nkhadka21) [reply slsim/image_simulation.py:309]: We just want to produce a empty image. I changed it little bit.

File: slsim/image_simulation.py (hunk 11 above)
Comment (nkhadka21) [reply slsim/image_simulation.py:557]: Image_ps contains list of all the individual lensed images of a source. So, adding them gives us a all lensed images in one frame.

File: slsim/Sources/point_plus_extended_source.py (hunk 1 above)
Comment (nkhadka21) [reply slsim/Sources/point_plus_extended_source.py:8]: Changed to PointPlusExtendedSource.

**Review by sibirrer — APPROVED** [review]
Looks great, thank you!

**Pull request merged**